programs/ssl/ssl_client2.c

```diff
@@ -716,6 +717,46 @@ static int ssl_save_session_serialize(mbedtls_ssl_context *ssl,
     return ret;
 }
 
+#if defined(MBEDTLS_SSL_EARLY_DATA)
+
+#define MBEDTLS_ERR_EARLY_FILE_IO_ERROR -2
+
+int ssl_early_data_read_file(const char *path, unsigned char **buffer, size_t *length)
```

Comment:
```suggestion
static int ssl_read_file(const char *path, unsigned char **buffer, size_t *length)
```
This is a generic function to read the content of a file into a buffer, I do not think we should restrict that to early data.

Reply:
> a generic function to read the content of a file into a buffer

It's a generic function to read a text file into a buffer, and null-terminate that buffer. While it could be used to read an arbitrary (binary) file, maybe it would be best to call it `ssl_read_file_text()` (i.e. read text data from a file)

Reply:
I do not think we should wrap it like this. It looks too complex and has `malloc` fail risk. 

```
FILE * fp=fopen(...)
....
n=fread(buf, 1, sizeof(buf), fp);
ssl_write_early_data(ssl,buf,n)) ;

exit:
   fclose(fp);
```

Reply:
Rename the function to ssl_read_file_text() done and for the time being, keep it be protected by macro guard MBEDTLS_SSL_EARLY_DATA.

Reply:
> I do not think we should wrap it like this. It looks too complex and has `malloc` fail risk.
> 
> ```
> FILE * fp=fopen(...)
> ....
> n=fread(buf, 1, sizeof(buf), fp);
> ssl_write_early_data(ssl,buf,n)) ;
> 
> exit:
>    fclose(fp);
> ```

If malloc failed, we think the test case should fail and stop, shouldn't it?  if so, we can avoid `malloc fail risk`.   And, cause we didn't know what size of the input file, will it be a better way to put `buf` on the stack?

Reply:
If `malloc()` is going to fail, either the file is huge or we have real problems. This is a program used by tests, not a part of the library, so let's keep using `malloc()`.

Reply:
`max_early_data_size` is up to `UINT_MAX` . It might cause `malloc()`.

Reply:
> `max_early_data_size` is up to `UINT_MAX`

But that is too much for the stack also!

Reply:
I think the plan is rather to read the file in chunks rather than in one go and store it in an allocated buffer. I think this is more aligned with real life thus I am rather in favor of the read by chunks option.

Reply:
Have revert the one go reading, and instead, will read it in chunks.